kv/kvclient/kvcoord: TestEvictionTokenCoalesce failed

kv/kvclient/kvcoord.TestEvictionTokenCoalesce [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7596404?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7596404?buildTab=artifacts#/) on master @ [8357abb668a5adaff781343b394b162fb1b66c6e](https://github.com/cockroachdb/cockroach/commits/8357abb668a5adaff781343b394b162fb1b66c6e):


```
=== RUN   TestEvictionTokenCoalesce
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestEvictionTokenCoalesce3398992622
    test_log_scope.go:79: use -show-logs to present logs inline
    dist_sender_test.go:4375: put encountered error: operation "range lookup" timed out after 45.119s (given timeout 10s): expected 2 in-flight requests, got 1
    dist_sender_test.go:4382: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestEvictionTokenCoalesce3398992622
--- FAIL: TestEvictionTokenCoalesce (45.20s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestEvictionTokenCoalesce.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-21629

kv/kvclient/kvcoord.TestEvictionTokenCoalesce [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7639370?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7639370?buildTab=artifacts#/) on master @ [dca415eddac0d659ae6d76b4e3dfdf4076adbd34](https://github.com/cockroachdb/cockroach/commits/dca415eddac0d659ae6d76b4e3dfdf4076adbd34):


```
=== RUN   TestEvictionTokenCoalesce
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestEvictionTokenCoalesce518070034
    test_log_scope.go:79: use -show-logs to present logs inline
    dist_sender_test.go:4375: put encountered error: operation "range lookup" timed out after 45.126s (given timeout 10s): expected 2 in-flight requests, got 1
    dist_sender_test.go:4382: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestEvictionTokenCoalesce518070034
--- FAIL: TestEvictionTokenCoalesce (45.23s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestEvictionTokenCoalesce.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvclient/kvcoord.TestEvictionTokenCoalesce [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7675767?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7675767?buildTab=artifacts#/) on master @ [39250f643837156c8fb75b0ced07ef8a99c14185](https://github.com/cockroachdb/cockroach/commits/39250f643837156c8fb75b0ced07ef8a99c14185):


```
=== RUN   TestEvictionTokenCoalesce
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestEvictionTokenCoalesce2090153120
    test_log_scope.go:79: use -show-logs to present logs inline
    dist_sender_test.go:4375: put encountered error: operation "range lookup" timed out after 45.129s (given timeout 10s): expected 2 in-flight requests, got 1
    dist_sender_test.go:4382: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestEvictionTokenCoalesce2090153120
--- FAIL: TestEvictionTokenCoalesce (45.30s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestEvictionTokenCoalesce.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvclient/kvcoord.TestEvictionTokenCoalesce [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7688345?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7688345?buildTab=artifacts#/) on master @ [b5be006bedd7d3cedc3fb3d2248df168e3d64be2](https://github.com/cockroachdb/cockroach/commits/b5be006bedd7d3cedc3fb3d2248df168e3d64be2):


```
=== RUN   TestEvictionTokenCoalesce
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestEvictionTokenCoalesce3954189781
    test_log_scope.go:79: use -show-logs to present logs inline
    dist_sender_test.go:4375: put encountered error: operation "range lookup" timed out after 45.132s (given timeout 10s): expected 2 in-flight requests, got 1
    dist_sender_test.go:4382: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestEvictionTokenCoalesce3954189781
--- FAIL: TestEvictionTokenCoalesce (45.25s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestEvictionTokenCoalesce.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvclient/kvcoord.TestEvictionTokenCoalesce [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7723480?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7723480?buildTab=artifacts#/) on master @ [1a6e9f885baa124d5ff2996adb966ea15a1a9b2b](https://github.com/cockroachdb/cockroach/commits/1a6e9f885baa124d5ff2996adb966ea15a1a9b2b):


```
=== RUN   TestEvictionTokenCoalesce
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestEvictionTokenCoalesce3402991081
    test_log_scope.go:79: use -show-logs to present logs inline
    dist_sender_test.go:4375: put encountered error: operation "range lookup" timed out after 45.114s (given timeout 10s): expected 2 in-flight requests, got 1
    dist_sender_test.go:4382: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestEvictionTokenCoalesce3402991081
--- FAIL: TestEvictionTokenCoalesce (45.19s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestEvictionTokenCoalesce.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvclient/kvcoord.TestEvictionTokenCoalesce [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7723484?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7723484?buildTab=artifacts#/) on master @ [1a6e9f885baa124d5ff2996adb966ea15a1a9b2b](https://github.com/cockroachdb/cockroach/commits/1a6e9f885baa124d5ff2996adb966ea15a1a9b2b):


```
=== RUN   TestEvictionTokenCoalesce
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestEvictionTokenCoalesce2253418981
    test_log_scope.go:79: use -show-logs to present logs inline
    dist_sender_test.go:4375: put encountered error: operation "range lookup" timed out after 45.122s (given timeout 10s): expected 2 in-flight requests, got 1
    dist_sender_test.go:4382: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestEvictionTokenCoalesce2253418981
--- FAIL: TestEvictionTokenCoalesce (45.26s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestEvictionTokenCoalesce.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvclient/kvcoord.TestEvictionTokenCoalesce [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7745807?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7745807?buildTab=artifacts#/) on master @ [2b595c5214aead3e7d4f47f9a1dedce57dba6e7f](https://github.com/cockroachdb/cockroach/commits/2b595c5214aead3e7d4f47f9a1dedce57dba6e7f):


```
=== RUN   TestEvictionTokenCoalesce
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestEvictionTokenCoalesce590097734
    test_log_scope.go:79: use -show-logs to present logs inline
    dist_sender_test.go:4375: put encountered error: operation "range lookup" timed out after 45.125s (given timeout 10s): expected 2 in-flight requests, got 1
    dist_sender_test.go:4382: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/40ec041d13327ceafaefe721fe51f834/logTestEvictionTokenCoalesce590097734
--- FAIL: TestEvictionTokenCoalesce (45.26s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestEvictionTokenCoalesce.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvclient/kvcoord/dist_sender_test.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender_test.go
@@ -4286,9 +4286,27 @@ func TestEvictionTokenCoalesce(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	sendErrors := int32(0)
-	var queriedMetaKeys sync.Map
+	// makeBarrier will make a function which will return once N goroutines
+	// have called it simultaneously. The first time these goroutines call
+	// this function, it will return false. After it has returned false,
+	// subsequent calls will return true.
+	makeBarrier := func(n int) func() (previouslyJoined bool) {
+		wg, done := sync.WaitGroup{}, atomic.Bool{}
+		wg.Add(n)
+		return func() bool {
+			if done.Load() {
+				return true
+			}
+			wg.Done()
+			wg.Wait()
+			done.Store(true)
+			return false
+		}
+	}
 
+	waitForInitialPuts := makeBarrier(2)
+	waitForInitialMeta2Scans := makeBarrier(2)
+	var queriedMetaKeys sync.Map
 	var ds *DistSender
 	testFn := func(ctx context.Context, ba *roachpb.BatchRequest) (*roachpb.BatchResponse, error) {
 		rs, err := keys.Range(ba.Requests)
@@ -4300,7 +4318,7 @@ func TestEvictionTokenCoalesce(t *testing.T) {
 		if !kv.TestingIsRangeLookup(ba) {
 			// Return a sendError so DistSender retries the first range lookup in the
 			// user key-space for both batches.
-			if atomic.AddInt32(&sendErrors, 1) <= 2 {
+			if previouslyWaited := waitForInitialPuts(); !previouslyWaited {
 				return nil, newSendError("boom")
 			}
 			return br, nil
@@ -4319,6 +4337,7 @@ func TestEvictionTokenCoalesce(t *testing.T) {
 			br.Add(r)
 			return br, nil
 		}
+		waitForInitialMeta2Scans()
 		// Querying meta2 range.
 		br = &roachpb.BatchResponse{}
 		r := &roachpb.ScanResponse{}

